README.md
# zekr
Zekr is an Open Source Qur'an study project.


doc/changes.txt
Version 0.1.0 beta 1 - [25-3-05]
Version 0.1.0 beta 2 - [29-3-05]

Version 0.1.0 release - [13-9-05] (First stable release)
Features/Enhancements:
	Nullsoft Installer (multilingual) support
	Basic navigation over suras and ayas
	.exe launcher for windows systems
	Sura ayas view layout: line-by-line, block
	Display sura info (descent type, aya count, sura number, etc.)
	Cache visited suras in a local directory for faster retrieval
Bug fixes:
	Problem with run.bat (JAVA_HOME) was fixed.
	Enter (Return) will work in Quran Form. (sura/aya selection)
	Localized hint for juz numbers
	Splash screen has title as well as icons.

Version 0.2.0 b1 - [9-12-05]
Features/Enhancements:
	Print a Quran view
	Search the whole Quran and also current view (a search result or a sura) with matching and non-matching diacritics
	Simple search statistics.
	Help menu: Zekr homepage, About.
	Language switch menu item for changing UI language while application is running through view menu.
	Configuration (layout, language, ...) is now saved in the xml file
Bug fixed:
	All misspelling for the wrong term sujda was changed to sajda.

Version 0.2.0 b2 - [24-12-05]
Features/Enhancements:
	Simple Save as menu for exporting a single sura or a search result
	Five language packs: Farsi (by Zekr team), Arabic (by Younes Moumen), English, French (by Younes Moumen), Indonesian (by Ari R. Fikri).
	Aya navigation enhancement: choosing to show an aya will also displays some previous ayas.
	Country flag icon for each language

Version 0.2.0 release - [20-1-06]
Features/Enhancements:
	Added German language pack (translated by Riyad Preukschas)
	Hyperlink for direct navigating to search result location (sura, aya) in the Quran
	More smoother Bismillah image.
	Enhanced presentation for sajda sign (character \u6E9)
	Text box replace with an editable combo box (can hold up to 40 search history).
	A Linux port is available (Mozilla 1.4+ GTK2 must be installed)
	A Mac port is available. (25-2-06)
Bug fixes:
	Unicode ALEF_MADDA (0x622) is now acted like alef when searching with non-matching diacritics
	All the following misspelling was changed from column one to column two:
	soora		sura
	joz			juz
	Search problem. ==> (أأنذرتهم was found but أنذرتهم could not be found)

Version 0.3.0 b1 - [14-3-06]
Features/Enhancements:
	Log enhancements. More info about system (OS, JRE, locale, ...) are stored.
	About menu enhancement.
	Menu items: Translation and Theme under View menu.
	Translation support:
		Basic support for translation packs.
		2 Different views for translation: 2 frames and inline.
		Ayatollah Makarem Shirazi's translation is added under a non-commercial use license
		A number of enhancements for BIDI support for translations view
	Theme support:
		A basic default theme (sky) is available.
		All themes are read at startup from [zekr]/res/theme, so every folder there is a potential theme!
		All themes should have a "theme.properties" file inside their folder (theme folder name is equal to theme ID).
		Does not have a GUI for setting options.
	InitRuntime changed to ApplicationRuntime, and is loaded once and can be accessed from ApplicationConfig.
	Translation layout (block/line-by-line) can be set as well as Quran layout.
	Tools>Add>Translation and Tools>Add>Themes. two menu items for importing translation and theme pack.
	M.H.Shakir's English translation will be available in the standard Zekr release under a non-commercial license.
	Apache Jakarta commons are stored in a single jar file, inside [zekr]/lib. Extra unused classes are removed from the jar file.
	Last seen Quran location is stored in zekr-config.xml, so that the next time Zekr starts, go to that location.
Bug fixes:
	JDK 1.4.2's JAXP SAX does not support BOM, so all language pack BOMs were removed.

Version 0.3.0 b2 - [25-3-06]
Features/Enhancements:
	Show popup translation for search result.
	4 different layouts:
		1. Quran only layout. show Quran only. Full Quran search is allowed in this view.
		2. Translation only layout. Full Quran search is not allowed, but one can search the current view.
		3. Separate; Quran at the top, translation at the bottom in 2 separate frames. Search is fully supported.
		4. Mixed; Quran and translation are mixed aya by aya. Each aya will be followed by its translation. Search is fully supported.
	Consequently, 4 new menu items under view>layout.
	Simple Options form under tools>options, with 2 different tabs:
		1. General tab: currently one option for enable/disable splash screen.
		2. View tab: a grid of key-value pairs for current theme. CRUD (Create, Read, Update and Delete) is fully supported on the grid.
Bug fixes:
	ALEF_MAKSURA will be treated as ARABIC_YEH when diacritic match is not set. In fact all YEH styled characters 
			(ARABIC_YEH, FARSI_YEH and ALEF_MAKSURA) both in Quran and search box are converted to ARABIC_YEH.
	A fine typo in Shakir translation was fixed.

Version 0.3.0 release - [25-5-06]
Features/Enhancements:
	All paths are now passed as URIs (instead of File.getAbsolutioPath()) to Velocity.
	Flag icons updated to more nice looking ones.
	Added Russian language pack (translated by Michael Kalinin).
	3 Russian translations (Porokhova, Osmanov, and Kuliev) provided by Michael Kalinin (under a non-commercial license). Will be as translation pack in the website (not present in the default distribution).
	The Quran text file completely changed from the old one to a more accurate one (after finding two typos in the old text file)
	View>Language moved to Tools>Options form, and now has a more nice look.
	View>Theme moved to Tools>Options form.
	Fooladvand (Farsi) translation will be available under a non-commercial license.
	A new sample theme (id: sample) is available.
	Max search result (default: 200) can be set on Options form.
	Updated to SWT build 3232.

zekr.sh
#!/bin/sh

# a native browser, mozilla or webkit, should be automatically looker up at runtime.
# if not found, you can can uncomment the following lines and point to a valid
# Firefox or XULRnner installation. see for more info: http://www.eclipse.org/swt/faq.php#specifyxulrunner

#export ZEKR_BROWSER="-Dorg.eclipse.swt.browser.XULRunnerPath=/usr/lib/firefox"

JAVA_CMD=java

ORIG_DIR_NAME=`cd`
DIR_NAME=`dirname $0`
MAIN_CLASS=net.sf.zekr.ZekrMain
CLASS_PATH=lib/log4j-1.2.8.jar:lib/swt.jar:lib/commons-collections-3.2.1.jar:lib/commons-codec-1.3.jar:lib/commons-io-1.4.jar:lib/commons-lang-2.4.jar:lib/commons-logging-1.0.4.jar:lib/commons-configuration-1.6.jar:lib/velocity-1.6.2.jar:lib/lucene-core-3.0.0.jar:lib/lucene-highlighter-3.0.0.jar:lib/lucene-snowball-3.0.0.jar:lib/lucene-memory-3.0.0.jar:lib/lucene-misc-3.0.0.jar:lib/mp3spi-1.9.4.jar:lib/vorbisspi-1.0.3.jar:lib/jlayer-1.0.1.jar:lib/basicplayer-3.0.jar:lib/tritonus-share-0.3.6.jar:lib/tritonus-jorbis-0.3.6.jar:lib/jorbis-0.0.17.jar:lib/jspeex-0.9.7.jar:dist/zekr.jar
VM_ARGS="-Xms20m -Xmx128m $ZEKR_BROWSER"

cd $DIR_NAME
"$JAVA_CMD" $VM_ARGS -cp "$CLASS_PATH" $MAIN_CLASS $*
cd $ORIG_DIR_NAME


res/ui/js/nav.js
/*
 *               In the name of Allah
 * This file is part of The Zekr Project. Use is subject to
 * license terms.
 *
 * @author Mohsen Saboorian
 */

var oldAyaId = null;

$(document).ready(function() {
	// backspace: history.back()
	if (!$.browser.msie) {
		$(document).keyup(function(e) {
			if (e.keyCode == 8) {
				var inp = e.target;
				if ("INPUT" == inp.nodeName.toUpperCase() && inp.type 
					&& "TEXT" == inp.type.toUpperCase())
					return; // by-pass this event
				history.go(-1);
			}
		});
	}

	/*$(window).resize(function(e) {
		refocus();
	});*/

	// $("#suraNav").focus();

    $('#quranSection a,#quranSection span').tipsy({fade: false, gravity: $.fn.tipsy.autoNS, opacity: 0.7});
   
    $(".commentHandle").click(function() {
   	    $(this).next('.commentText:eq(0)').toggle(); 
   	});
});

function refocus() {
	var suraNum = $("input#hiddenSuraNum").val();
	var ayaNum = $("input#hiddenAyaNum").val();

	var ayaId = suraNum + "_" + ayaNum;	
	var aya = document.getElementById(ayaId);
	if (!aya) return;

	if (ayaNum != 1)
		$(aya).ScrollTo(1, 'original', getBrowserHeight() > getObjectHeight(aya) ? getBrowserHeight()/5 : 0);
	else
		$(aya).ScrollTo(1, 'original', 200);

}

function navtoSuraAya() {
	var sura = $("input#suraNav").val();
	var aya = $("input#ayaNav").val();
	var page = $("input#pageNav").val();
	var origSuraNum = $("input#hiddenSuraNum").val();
	if (!isNaN(parseInt(sura.trim())) && !isNaN(parseInt(aya.trim())))
		if (origSuraNum != sura)
			gotoSuraAya(sura, aya, page);
		else
			gotoSuraAya(sura, aya, page);
}

function getObjectHeight(obj) { return obj.offsetHeight; }
function getBrowserHeight() { return document.body.clientHeight; }

function highlightAya(id) {
	$('#sign_' + id).addClass('selectedAyaSign');
	$('#' + id).addClass('selectedAya');
}

function unHighlightAya(id) {
    $('#sign_' + id).removeClass('selectedAyaSign');
    $('#' + id).removeClass('selectedAya');
}

function focusOnAya(suraNum, ayaNum) {
	var ayaId = suraNum + "_" + ayaNum;	
	var aya = document.getElementById(ayaId);
	if (!aya) return;

	$("input#suraNav").val(suraNum);
	$("input#ayaNav").val(ayaNum);
	$("input#hiddenSuraNum").val(suraNum);
	$("input#hiddenAyaNum").val(ayaNum);

	if (oldAyaId != null)
		unHighlightAya(oldAyaId);
	highlightAya(ayaId);

	var bh = $(window).height();

	// var x = $(aya).position().top;
	// $("html,body").animate({scrollTop: ayaNum == 1 ? x - 200 : x - bh / 4}, 400, 'easeOutQuad'); //easeInCubic
	var t = $(aya).offset().top;
	$("html,body").animate({scrollTop: ayaNum == 1 ? t - 150 : t - bh / 4}, 400, 'easeOutQuad');

	oldAyaId = ayaId;
}


res/ui/js/common.js
/*
 *               In the name of Allah
 * This file is part of The Zekr Project. Use is subject to
 * license terms.
 *
 * @author Mohsen Saboorian
 */

String.prototype.trim = function() {
    return this.replace(/([\s\xA0]+$)|(^[\s\xA0]+)/g, '');
};

var ayaFocusHooks = [];

// Browser <-> SWT communication
function gotoSuraAya(sura, aya, page) { // use it when sura changed
    if (javaFunction) {
        javaFunction('ZEKR::GOTO', sura, aya, page);
    }
}
function redirect(sura, aya) { // used for search results
    if (javaFunction) {
        javaFunction('ZEKR::REDIRECT', sura, aya);
    }
}
function translate(sura, aya) {
    if (javaFunction) {
        javaFunction('ZEKR::TRANS', sura, aya);
    }
}
function play(loc, isPlay) {
    if (javaFunction) {
        javaFunction('ZEKR::PLAY', loc, isPlay);
    }
}

function zoom(z, trans) {
    if (javaFunction) {
        javaFunction('ZEKR::ZOOM', z, trans);
    }
}

function error(e) {
	alert("An unexpected error occurred:\n" + "[" + e.name + ":" + e.number + "] " + e.message + "\n");
}


// tipsy, facebook style tooltips for jquery
// version 1.0.0a
// (c) 2008-2010 jason frame [jason@onehackoranother.com]
// released under the MIT license

(function($) {
    
    function maybeCall(thing, ctx) {
        return (typeof thing == 'function') ? (thing.call(ctx)) : thing;
    };
    
    function isElementInDOM(ele) {
      while (ele = ele.parentNode) {
        if (ele == document) return true;
      }
      return false;
    };
    
    function Tipsy(element, options) {
        this.$element = $(element);
        this.options = options;
        this.enabled = true;
        this.fixTitle();
    };
    
    Tipsy.prototype = {
        show: function() {
            var title = this.getTitle();
            if (title && this.enabled) {
                var $tip = this.tip();
                
                $tip.find('.tipsy-inner')[this.options.html ? 'html' : 'text'](title);
                $tip[0].className = 'tipsy'; // reset classname in case of dynamic gravity
                $tip.remove().css({top: 0, left: 0, visibility: 'hidden', display: 'block'}).prependTo(document.body);
                
                var pos = $.extend({}, this.$element.offset(), {
                    width: this.$element[0].offsetWidth,
                    height: this.$element[0].offsetHeight
                });
                
                var actualWidth = $tip[0].offsetWidth,
                    actualHeight = $tip[0].offsetHeight,
                    gravity = maybeCall(this.options.gravity, this.$element[0]);
                
                var tp;
                switch (gravity.charAt(0)) {
                    case 'n':
                        tp = {top: pos.top + pos.height + this.options.offset, left: pos.left + pos.width / 2 - actualWidth / 2};
                        break;
                    case 's':
                        tp = {top: pos.top - actualHeight - this.options.offset, left: pos.left + pos.width / 2 - actualWidth / 2};
                        break;
                    case 'e':
                        tp = {top: pos.top + pos.height / 2 - actualHeight / 2, left: pos.left - actualWidth - this.options.offset};
                        break;
                    case 'w':
                        tp = {top: pos.top + pos.height / 2 - actualHeight / 2, left: pos.left + pos.width + this.options.offset};
                        break;
                }
                
                if (gravity.length == 2) {
                    if (gravity.charAt(1) == 'w') {
                        tp.left = pos.left + pos.width / 2 - 15;
                    } else {
                        tp.left = pos.left + pos.width / 2 - actualWidth + 15;
                    }
                }
                
                $tip.css(tp).addClass('tipsy-' + gravity);
                $tip.find('.tipsy-arrow')[0].className = 'tipsy-arrow tipsy-arrow-' + gravity.charAt(0);
                if (this.options.className) {
                    $tip.addClass(maybeCall(this.options.className, this.$element[0]));
                }
                
                if (this.options.fade) {
                    $tip.stop().css({opacity: 0, display: 'block', visibility: 'visible'}).animate({opacity: this.options.opacity});
                } else {
                    $tip.css({visibility: 'visible', opacity: this.options.opacity});
                }
            }
        },
        
        hide: function() {
            if (this.options.fade) {
                this.tip().stop().fadeOut(function() { $(this).remove(); });
            } else {
                this.tip().remove();
            }
        },
        
        fixTitle: function() {
            var $e = this.$element;
            if ($e.attr('title') || typeof($e.attr('original-title')) != 'string') {
                $e.attr('original-title', $e.attr('title') || '').removeAttr('title');
            }
        },
        
        getTitle: function() {
            var title, $e = this.$element, o = this.options;
            this.fixTitle();
            var title, o = this.options;
            if (typeof o.title == 'string') {
                title = $e.attr(o.title == 'title' ? 'original-title' : o.title);
            } else if (typeof o.title == 'function') {
                title = o.title.call($e[0]);
            }
            title = ('' + title).replace(/(^\s*|\s*$)/, "");
            return title || o.fallback;
        },
        
        tip: function() {
            if (!this.$tip) {
                this.$tip = $('<div class="tipsy"></div>').html('<div class="tipsy-arrow"></div><div class="tipsy-inner"></div>');
                this.$tip.data('tipsy-pointee', this.$element[0]);
            }
            return this.$tip;
        },
        
        validate: function() {
            if (!this.$element[0].parentNode) {
                this.hide();
                this.$element = null;
                this.options = null;
            }
        },
        
        enable: function() { this.enabled = true; },
        disable: function() { this.enabled = false; },
        toggleEnabled: function() { this.enabled = !this.enabled; }
    };
    
    $.fn.tipsy = function(options) {
        
        if (options === true) {
            return this.data('tipsy');
        } else if (typeof options == 'string') {
            var tipsy = this.data('tipsy');
            if (tipsy) tipsy[options]();
            return this;
        }
        
        options = $.extend({}, $.fn.tipsy.defaults, options);
        
        function get(ele) {
            var tipsy = $.data(ele, 'tipsy');
            if (!tipsy) {
                tipsy = new Tipsy(ele, $.fn.tipsy.elementOptions(ele, options));
                $.data(ele, 'tipsy', tipsy);
            }
            return tipsy;
        }
        
        function enter() {
            var tipsy = get(this);
            tipsy.hoverState = 'in';
            if (options.delayIn == 0) {
                tipsy.show();
            } else {
                tipsy.fixTitle();
                setTimeout(function() { if (tipsy.hoverState == 'in') tipsy.show(); }, options.delayIn);
            }
        };
        
        function leave() {
            var tipsy = get(this);
            tipsy.hoverState = 'out';
            if (options.delayOut == 0) {
                tipsy.hide();
            } else {
                setTimeout(function() { if (tipsy.hoverState == 'out') tipsy.hide(); }, options.delayOut);
            }
        };
        
        if (!options.live) this.each(function() { get(this); });
        
        if (options.trigger != 'manual') {
            var binder   = options.live ? 'live' : 'bind',
                eventIn  = options.trigger == 'hover' ? 'mouseenter' : 'focus',
                eventOut = options.trigger == 'hover' ? 'mouseleave' : 'blur';
            this[binder](eventIn, enter)[binder](eventOut, leave);
        }
        
        return this;
        
    };
    
    $.fn.tipsy.defaults = {
        className: null,
        delayIn: 0,
        delayOut: 0,
        fade: false,
        fallback: '',
        gravity: 'n',
        html: false,
        live: false,
        offset: 0,
        opacity: 0.8,
        title: 'title',
        trigger: 'hover'
    };
    
    $.fn.tipsy.revalidate = function() {
      $('.tipsy').each(function() {
        var pointee = $.data(this, 'tipsy-pointee');
        if (!pointee || !isElementInDOM(pointee)) {
          $(this).remove();
        }
      });
    };
    
    // Overwrite this method to provide options on a per-element basis.
    // For example, you could store the gravity in a 'tipsy-gravity' attribute:
    // return $.extend({}, options, {gravity: $(ele).attr('tipsy-gravity') || 'n' });
    // (remember - do not modify 'options' in place!)
    $.fn.tipsy.elementOptions = function(ele, options) {
        return $.metadata ? $.extend({}, options, $(ele).metadata()) : options;
    };
    
    $.fn.tipsy.autoNS = function() {
        return $(this).offset().top > ($(document).scrollTop() + $(window).height() / 2) ? 's' : 'n';
    };
    
    $.fn.tipsy.autoWE = function() {
        return $(this).offset().left > ($(document).scrollLeft() + $(window).width() / 2) ? 'e' : 'w';
    };
    
    /**
     * yields a closure of the supplied parameters, producing a function that takes
     * no arguments and is suitable for use as an autogravity function like so:
     *
     * @param margin (int) - distance from the viewable region edge that an
     *        element should be before setting its tooltip's gravity to be away
     *        from that edge.
     * @param prefer (string, e.g. 'n', 'sw', 'w') - the direction to prefer
     *        if there are no viewable region edges effecting the tooltip's
     *        gravity. It will try to vary from this minimally, for example,
     *        if 'sw' is preferred and an element is near the right viewable 
     *        region edge, but not the top edge, it will set the gravity for
     *        that element's tooltip to be 'se', preserving the southern
     *        component.
     */
     $.fn.tipsy.autoBounds = function(margin, prefer) {
		return function() {
			var dir = {ns: prefer[0], ew: (prefer.length > 1 ? prefer[1] : false)},
			    boundTop = $(document).scrollTop() + margin,
			    boundLeft = $(document).scrollLeft() + margin,
			    $this = $(this);

			if ($this.offset().top < boundTop) dir.ns = 'n';
			if ($this.offset().left < boundLeft) dir.ew = 'w';
			if ($(window).width() + $(document).scrollLeft() - $this.offset().left < margin) dir.ew = 'e';
			if ($(window).height() + $(document).scrollTop() - $this.offset().top < margin) dir.ns = 's';

			return dir.ns + (dir.ew ? dir.ew : '');
		}
	};
    
})(jQuery);


res/ui/js/search.js
/**
 *               In the name of Allah
 * This file is part of The Zekr Project. Use is subject to
 * license terms.
 *
 * @author Mohsen Saboorian
 */

var SUKUN = String.fromCharCode(0x652);
var SHADDA = String.fromCharCode(0x651);
var KASRA = String.fromCharCode(0x650);
var DAMMA = String.fromCharCode(0x64f);
var FATHA = String.fromCharCode(0x64e);

var KASRATAN = String.fromCharCode(0x64d);
var DAMMATAN = String.fromCharCode(0x64c);
var FATHATAN = String.fromCharCode(0x64b);

var SUPERSCRIPT_ALEF = String.fromCharCode(0x670);

var HAMZA = String.fromCharCode(0x621);
var ALEF = String.fromCharCode(0x627);
var ALEF_MADDA = String.fromCharCode(0x622);
var ALEF_HAMZA_ABOVE = String.fromCharCode(0x623);
var ALEF_HAMZA_BELOW = String.fromCharCode(0x625);

var YEH_HAMZA_ABOVE = String.fromCharCode(0x626);
var WAW_HAMZA_ABOVE = String.fromCharCode(0x624);
var WAW = String.fromCharCode(0x648);

var ALEF_MAKSURA = String.fromCharCode(0x649);
var FARSI_YEH = String.fromCharCode(0x6cc);
var ARABIC_YEH = String.fromCharCode(0x64a);

var ARABIC_KAF = String.fromCharCode(0x643);
var FARSI_KEHEH = String.fromCharCode(0x6a9);
	
function replaceAll(str, oldStr, newStr) {
	var i = str.indexOf(oldStr);
	var newLen = newStr.length;
	while (i > -1) {
		str = str.replace(oldStr, newStr);
		i = str.indexOf(oldStr, i + newLen);
	}
	return str;
}

function replaceSimilarArabic(str) {
	str = replaceAll(str, ALEF_MAKSURA, ARABIC_YEH);
	str = replaceAll(str, ALEF_HAMZA_ABOVE, ALEF);
	str = replaceAll(str, ALEF_HAMZA_BELOW, ALEF);
	str = replaceAll(str, ALEF_MADDA, ALEF);
	return str;
}

function arabicSimplify(str) {
	// diacritics removal
	var arr = [SUKUN, SHADDA, KASRA, DAMMA, FATHA, KASRATAN, DAMMATAN, FATHATAN, SUPERSCRIPT_ALEF];
	for (var i = 0; i < arr.length; i++) {
		str = replaceAll(str, arr[i], "");
	}

	// Some replacement
	str = replaceAll(str, ALEF_MAKSURA, ARABIC_YEH);
	str = replaceAll(str, FARSI_YEH, ARABIC_YEH);
	str = replaceAll(str, FARSI_KEHEH, ARABIC_KAF);
	str = replaceAll(str, ALEF_HAMZA_ABOVE, ALEF);
	str = replaceAll(str, ALEF_HAMZA_BELOW, ALEF);
	str = replaceAll(str, ALEF_MADDA, ALEF);
	return str;
}

function isDiac(ch) {
	return (ch == SUKUN) || (ch == SHADDA) || (ch == KASRA) || (ch == DAMMA) || 
	       (ch == FATHA) || (ch == KASRATAN) || (ch == DAMMATAN) || (ch == FATHATAN) || 
	       (ch == SUPERSCRIPT_ALEF);
}

function indexOfIgnoreDiacritic(src, key) {
	key = arabicSimplify(key);
	src = replaceSimilarArabic(src);
	var k = 0, s = 0, start = -1;
	if (key.length == 0)
		return -1;
	while(s < src.length) {
		if (k == key.length)
			break;

		if (src.charAt(s) == key.charAt(k)) {
			if (start == -1)
				start = s;
			s++; k++;
		} else {
			if (!isDiac(src.charAt(s))) {
				if (k != 0)
					s--;
				k = 0;
				start = -1;
			}
			s++;
		}
	}
	if (k == key.length) { // fully matched
		spaceBefore = (key.charAt(0) != ' ') ? src.substring(0, start).lastIndexOf(' ') : start;
		spaceAfter = (key.charAt(key.length - 1) != ' ') ? src.indexOf(' ', s) : s;
		if (spaceBefore == -1) start = 0;
		else start = spaceBefore + 1;
		if (spaceAfter == -1) s = src.length;
		else s = spaceAfter;
		return {startIndex: start, endIndex: s};
	}
	return -1;
}

function indexOfMatchDiacritic(src, key) {
	start = src.indexOf(key);
	if (start == -1)
		return -1;

	spaceBefore = (key.charAt(0) != ' ') ? src.substring(0, start).lastIndexOf(' ') : start;
	spaceAfter = (key.charAt(key.length - 1) != ' ') ? src.indexOf(' ', start + key.length) : start + key.length;
	if (spaceBefore == -1) start = 0;
	else start = spaceBefore + 1;
	if (spaceAfter == -1) end = src.length;
	else end = spaceAfter;
	return {startIndex: start, endIndex: end};
}

function highlightWordInNode(aWord, aNode, matchDiac, matchCase) {
	if (aNode.nodeType == 1){
		var children = aNode.childNodes;
		for(var i = 0; i < children.length; i++) {
			highlightWordInNode(aWord, children[i], matchDiac, matchCase);
		}
    }
	else if (aNode.nodeType == 3){
		highlightWordInText(aWord, aNode, matchDiac, matchCase);
	}
}

function highlightWordInText(aWord, textNode, matchDiac, matchCase){
	var allText = new String(textNode.data);

	var lower = ""
	if (!matchCase) { lower = allText.toLowerCase(); aWord = aWord.toLowerCase(); }
	else lower = allText;

	var myIndexOf;
	if (matchDiac)
		myIndexOf = indexOfMatchDiacritic;
	else
		myIndexOf = indexOfIgnoreDiacritic;

	loc = myIndexOf(lower, aWord);
	if (loc == -1) return;

	// create a node to replace the textNode so we end up
	// not changing number of children of textNode.parent
	replacementNode = document.createElement("span");
	textNode.parentNode.insertBefore(replacementNode, textNode);
	while (loc != -1){
		sIndex = loc.startIndex;
		eIndex = loc.endIndex;
		before = allText.substring(0, sIndex);
		newBefore = document.createTextNode(before);
		replacementNode.appendChild(newBefore);
		spanNode = document.createElement("span");
		spanNode.className = "jsHighlight";
		spanNode.title = aWord;
		replacementNode.appendChild(spanNode);
		boldText = document.createTextNode(allText.substring(sIndex, eIndex));
		spanNode.appendChild(boldText);
		allText = allText.substring(eIndex);
		lower = matchCase ? allText : allText.toLowerCase();
		loc = myIndexOf(lower, aWord);
	}
	newAfter = document.createTextNode(allText);
	replacementNode.appendChild(newAfter);
	textNode.parentNode.removeChild(textNode);
}

Finder = function(matchDiac, matchCase) {
	this.matchDiac = matchDiac;
	this.matchCase = matchCase;
}

Finder.prototype.find = function(str) {
	if (str == "") return;
	highlightWordInNode(str, document.getElementById("searchableSection"), this.matchDiac, this.matchCase);
}

find = function(str, matchDiac, matchCase) {
	if (matchCase === undefined) matchCase = false;

	$("span.jsHighlight").each(function() {
		var p = $(this).parent().eq(0);
		p.after(p.text()).remove();
	});
	$("span.jsHighlight").each(function() {
		$(this).after($(this).text()).remove();
	});
	$("span.jsHighlightFocused").each(function() {
		var p = $(this).parent().eq(0);
		p.after(p.text()).remove();
	});
	$("span.jsHighlightFocused").each(function() {
		$(this).after($(this).text()).remove();
	});

//	$("span.jsHighlightFocused").removeClass("jsHighlightFocused");
//	$("span.jsHighlight").removeClass("jsHighlight");

	var finder = new Finder(matchDiac, matchCase);
	finder.find(str);

	// reveal next-prev buttons (if possible: this is not possible in a whole-quran search view)
	$("#nextPrevButtons").show("", new function() {
		$("#nextPrevButtons")[0]._res = new CurrentPageSearchResult();
	});
};


SearchResult = function() {
	var cnt;
	var num = 0;
	var oldNum = 0;
	var list;
	$(document).ready(function() {
		try{
			cnt = $("div.searchResult .item").size();
			list = $("div.searchResult>div");
		} catch(e) {error(e); return;}
		focus();
	});

	this.next = function() {
		if (num < cnt - 1) { oldNum = num; num++; focus(); }
	};

	this.prev = function() {
		if (num > 0) { oldNum = num; num--; focus(); }
	};

	function focus() {
		if (cnt <= 0)
			return;

		// var bh = getBrowserHeight();
		var bh = $(window).height();
		$("#result_" + (1+oldNum)).children("div").attr("className", "item");
		var ch = $("#result_" + (1+num)).children("div");
		ch.attr("className", "selectedAya");
		var x = ch.offset().top;
		$("html,body").animate({scrollTop: x - bh / 4}, 400, 'easeOutQuad'); // easeInCubic

		var tt = $("#itemNum_" + (1+num)).attr("title");
		if (!tt) {
			tt = $("#itemNum_" + (1+num)).attr("original-title");
		}
		
		var suraAya = tt.split('-');
		$("#suraNum").val(suraAya[0]);
		$("#ayaNum").val(suraAya[1]);
		
	};
};

CurrentPageSearchResult = function() {
	var cnt;
	var num = 0;
	var oldNum = 0;
	var list;
	$(document).ready(function() {
		try{
			list = $("span.jsHighlight");
			cnt = list.size();
		} catch(e) {error(e); return;}
		if (cnt > 0)
			focus();
	});

	this.next = function() {
	alert(cnt);
		if (cnt <= 0) return;
		oldNum = num;
		num < cnt - 1 ? num++ : num = 0;
		focus();
	};

	this.prev = function() {
		if (cnt <= 0) return;
		oldNum = num;
		num > 0 ? num-- : num = cnt - 1; 
		focus(); 
	};

	function focus() {
		var h = list.eq(num).height();
		var bh = getBrowserHeight();
		var item = list.get(num);
		$("#focusedWord").html("\"" + $(item).text() + "\"");
		$(list.get(oldNum)).attr("className", "jsHighlight");
		$(item).ScrollTo(500, 'original', bh > h  ? bh/5 : 0).attr("className", "jsHighlightFocused");
	};
};

var lastPlayingSearchItem = false;
function searchResultRecite(but, intrnl) {
	var b = $(but);
	if (!intrnl && lastPlayingSearchItem && b.attr('id') != $(lastPlayingSearchItem).attr('id') && $(lastPlayingSearchItem).attr('playing')) {
		searchResultRecite(lastPlayingSearchItem, true);
	}

	lastPlayingSearchItem = but;
	var isPlay = !b.attr('playing');
	togglePlayControl(b.attr('loc'), isPlay);
	play(b.attr('loc'), '' + isPlay);
}

function togglePlayControl(loc, isPlay) {
	b = $('#item_' + loc);
	b.attr('playing', isPlay);
	var img = b.find('img');
	var i = img.attr('zekricon');
	var t = b.attr('zekrtitle');
	img.attr('zekricon', img.attr('src'));
	b.attr('zekrtitle', b.attr('title'));
	img.attr('src', i);
	b.attr('title', t);
}

function stopPlayingItem() {
	togglePlayControl($(lastPlayingSearchItem).attr('loc'), false);
}

function gotoFoundItem(sura, aya) {
	b = $(lastPlayingSearchItem);
	isPlay = b.attr('playing');
	if (isPlay) {
		play(b.attr('loc'), 'false');
		stopPlayingItem();
	}
	redirect(sura, aya);
}